internal/controllers/core/tiltfile/reconciler.go

```diff
@@ -151,6 +165,69 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 	return ctrl.Result{}, nil
 }
 
+// Modeled after BuildController.needsBuild and NextBuildReason(). Check to see that:
+// 1) There's currently no Tiltfile build running,
+// 2) There are pending file changes, and
+// 3) Those files have changed since the last Tiltfile build
+//    (so that we don't keep re-running a failed build)
+// 4) OR the command-line args have changed since the last Tiltfile build
+// 5) OR user has manually triggered a Tiltfile build
+//
+// In some cases, there may be race conditions where a tiltfile has finished executing,
+// but its results haven't bee
+func (r *Reconciler) needsBuild(ctx context.Context, nn types.NamespacedName, tf *v1alpha1.Tiltfile, run *runStatus, fileWatches map[string]*v1alpha1.FileWatch, triggerQueue *v1alpha1.ConfigMap, lastRestartEvent time.Time) *BuildEntry {
+	var reason model.BuildReason
+	filesChanged := []string{}
+
+	step := runStepNone
+	lastStartTime := time.Time{}
+	lastStartArgs := []string{}
+	if run != nil {
+		step = run.step
+		lastStartTime = run.startTime
+		lastStartArgs = run.startArgs
+	}
+
+	if step == runStepNone {
+		reason = reason.With(model.BuildReasonFlagInit)
+	} else {
+		filesChanged = restarton.FilesChanged(tf.Spec.RestartOn, fileWatches, lastStartTime)
+		if len(filesChanged) > 0 {
+			reason = reason.With(model.BuildReasonFlagChangedFiles)
+		} else if lastRestartEvent.After(lastStartTime) {
+			reason = reason.With(model.BuildReasonFlagTriggerUnknown)
```

Comment:
I'm guessing there's broader implications (e.g. handling in UI), but should we add some sort of `BuildReasonFlagTriggerExternal` as a catch-all in the future?

Reply:
Ah nevermind - I was thinking that "Unknown" was the default/0 value, i.e. would more indicative of a logic error that we forgot to populate it, but I see we already use(d) it as the umbrella, so it's really as you described it 👍